File: Daily challenge/Minimum Falling Path Sum II.cpp
class Solution 
{
public:
    int minFallingPathSum(vector<vector<int>>& grid) 
    {
        int n = grid.size();
        int m = grid[0].size();
        vector<int> dp = grid[0];

        for(int i = 1; i < n; i++)
        {
            vector<int> next_dp(m, INT_MAX);
            for(int j = 0; j < m; j++)
            {
                for(int prev = 0; prev < m; prev++)
                {
                    if(prev != j)
                    {
                        next_dp[j] = min(next_dp[j], grid[i][j] + dp[prev]);
                    }
                }
            }
            dp = next_dp;
        }
        return *min_element(dp.begin(), dp.end());
    }
};

File: Daily challenge/Minimum Falling Path Sum II.py
class Solution:
    def minFallingPathSum(self, grid: List[List[int]]) -> int:
        n = len(grid)
        m = len(grid[0])
        dp = grid[0] 

        for i in range(1, n):
            next_dp = [float('inf')] * n
            for col in range(m):
                for prev_col in range(m):
                    if prev_col != col:
                        next_dp[col] = min(next_dp[col], grid[i][col] + dp[prev_col])
            dp = next_dp
        return min(dp)